Backendwork

ivangoudyrev opened the pull request:
Completed models and views for Contact List functionalities of the following models: Contactlist, Client, Agent, Inspector, Lender, Title.  
Functionalities include ability to get lists and individual instances, specific to the logged-in user.  Additionally, the functionality to modify or delete specific entries.

Next close-range: back-end functionality to get/post/put/delete Task and Subtask entries.
Next far-range: front-end work.

ivangoudyrev committed:
continued work on backend urls and views

ivangoudyrev committed:
completed views for contactslist, client, agent, lender, title, and inspector

fravila08 commented on TransactionManager/back_end/agent_app/__pycache__/admin.cpython-310.pyc:
Please add `__pycache__` to a `.gitignore` file and ensure to remove these directories from your repository

ivangoudyrev replied:
Umar had me add these lines to `.gitignore`.  
```
__pycache__/
*.py[cod]
*$py.class
```
Should I do it for every app? Like this:
```
agent_app/__pycache__
client_app/__pycache__
[...]
```
Also, should I also add my migrations folders?  They seem to be accumulating a lot of extra files.

fravila08 replied:
You should have a .gitignore at the root of your project this will make sure that those files aren't being tracked. With that said when you make this change you'll have to manually delete these __pycache__ directories, commit and push the changes

ivangoudyrev replied:
done.  I made sure that the line `__pycache__/` is in the `.gitignore` file.  I also deleted every `__pycache__` directory in my app folders. However, they all re-appeared on the next `makemigrations `. Not sure if this worked though.  The new `__pycache__` folders are not greyed out...

fravila08 commented on TransactionManager/back_end/agent_app/serializers.py:
```diff
@@ -0,0 +1,8 @@
+from rest_framework.serializers import ModelSerializer
+from .models import Agent
+
+class AAgentSerializer(ModelSerializer):
```
I think you may have a typo here

ivangoudyrev replied:
Not a typo... I actually have a "A Agent Serializer".  'A' implies that it's for a singe instance of Agent class

fravila08 commented on TransactionManager/back_end/agent_app/views.py:
```diff
@@ -1,3 +1,37 @@
+        existing_agent = get_object_or_404(request.user.contacts.agents, id=id)
+        serializer = AAgentSerializer(existing_agent, data=request.data)    
+        if serializer.is_valid():
+            serializer.save()
```
Good error handling, I recommend you place some form of error boundary for all views to indicate to users that the problem is within their request

ivangoudyrev replied:
Thank you, I actually stole that bit of code.  Will work through the rest to clean up and make error handling more uniform.

ivangoudyrev replied:
I will add improved error handling on the next pull.  Need to think through edge cases and contingencies first.  Some of the views will be generated on-button-click and probably won't need to have a robust error handling.  Some of the potential errors are already prevented by the validators.  Anyway, please approve the pull without this one complete for now.

fravila08 commented on TransactionManager/back_end/client_app/serializers.py:
```diff
@@ -0,0 +1,8 @@
+    
+    class Meta:
+        model = Client
+        fields = ['id', 'first_name', 'mid_init', 'last_name', 'phone', 'email']
```
if these are all the fields within your model, you could save some time and space by using `fields = ["__all__"]

ivangoudyrev replied:
Done.  I will be using all fields in the front end.  Made the changes in all serializers.

ivangoudyrev replied:
FYI: `fields = ["all"]` didn't work.  `fields = '__all__'` worked

fravila08 commented on TransactionManager/back_end/client_app/views.py:
```diff
@@ -1,3 +1,37 @@
+from .models import Client
+from .serializers import AClientSerializer
+
+class User_permissions(APIView):
```
I believe you've already made this `User_permissions` class. Instead of creating it in every app it may be a good idea to import it. Don't forget your DRY principle

ivangoudyrev replied:
Thank you! I didn't know I could do that.  Will do that for all apps.  
What do you think about this: create a `permissions.py` file in user_app with just:
```
from rest_framework.views import APIView
class User_permissions(APIView):
    authentication_classes = [TokenAuthentication]
    permission_classes = [IsAuthenticated]
```
then, in every views file that requires it:
`from user_app.permissions import User_permissions`

ivangoudyrev replied:
Wilco.

ivangoudyrev replied:
Done.

fravila08 commented on TransactionManager/back_end/user_app/views.py:
```diff
@@ -1,3 +1,63 @@
+            "first_name": request.user.first_name, 
+            "last_name": request.user.last_name, 
+            "email": request.user.email
+        })
```
This is probably a good place to create that `User_permissions` class to import in other files

ivangoudyrev replied:
Will do. Thank you.

ivangoudyrev replied:
Done.

fravila08 reviewed (changes requested):
Great code base so far, please remove the `__pycache__` from your git history and repo. You also have a bit of repeated code in different portions of your view files. Please refactor this information. Maybe create helper functions in a `utilities.py` file to isolate these behaviors and then import them to your separate views files. Don't forget "Dry syrup"(DRY SRP) principles.

fravila08 commented:
Please push another commit to this branch so I can approve it

ivangoudyrev commented:
> Will do.  Making the changes you mentioned and will push another commit later today.  
Also, thank you for the detailed feedback!

ivangoudyrev committed:
Code clean up and attemped removal of pycache

ivangoudyrev commented:
-Moved the User_permissions function into user_app and imported throughout, while deleting replication.
-Added __pycache__ to .gitignore and deleted from all folders.  For some reason it's not greying out in my vscode.  I hope this pull removes it. 
-Replaced listing of all fields with fields = '__all__' where all fields are needed.
-Will improve error handling on the next pull.

Please approve...

ivangoudyrev committed:
completed models and views for all remaining apps

ivangoudyrev committed:
started frontend here by accident.  commit before switching branches

ivangoudyrev committed:
Merge branch 'main' into backendwork
